openwisp_monitoring/notifications/admin.py

```diff
@@ -57,11 +77,87 @@ def mark_as_read(self, request, queryset):
 
     mark_as_read.short_description = _('Mark selected notifications as read')
 
+    def actor_object_link(self, obj):
+        try:
+            url = reverse(
+                'admin:{0}_{1}_change'.format(
+                    obj.actor_content_type.app_label, obj.actor_content_type.model
+                ),
+                args=(obj.actor_object_id,),
+            )
+            return format_html(
+                '<a href="{url}" id="actor-object-url">{id}</a>',
+                url=url,
+                id=obj.actor_object_id,
+            )
+        except NoReverseMatch:
+            return obj.actor_object_id
+        except AttributeError:
+            return '-'
+
+    actor_object_link.short_description = _('Actor Object')
+
+    def action_object_link(self, obj):
+        try:
+            url = reverse(
+                'admin:{0}_{1}_change'.format(
+                    obj.action_object_content_type.app_label,
+                    obj.action_object_content_type.model,
+                ),
+                args=(obj.action_object_object_id,),
+            )
+            return format_html(
+                '<a href="{url}" id="action-object-url">{id}</a>',
+                url=url,
+                id=obj.action_object_object_id,
+            )
+        except NoReverseMatch:
+            return obj.action_object_object_id
+        except AttributeError:
+            return '-'
+
+    action_object_link.short_description = _('Action Object')
+
+    def target_object_link(self, obj):
+        try:
+            url = reverse(
+                'admin:{0}_{1}_change'.format(
+                    obj.target_content_type.app_label, obj.target_content_type.model
+                ),
+                args=(obj.target_object_id,),
+            )
+            return format_html(
+                '<a href="{url}" id="target-object-url">{id}</a>',
+                url=url,
+                id=obj.target_object_id,
+            )
+        except NoReverseMatch:
+            return obj.target_object_id
+        except AttributeError:
+            return '-'
```

Comment:
the previous 3 methods have identical structure so it's good to generalize its code and avoid repetition.

Try starting from here (I did not test this code):

```python
def _get_link(self, obj, field, html=True):
    content_type = getattr(obj, f'{field}_content_type')
    object_id = getattr(obj, f'{field}_object_id')
    try:
        url = reverse(
            f'admin:{content_type.app_label}_{content_type.model}_change'
            args=[object_id],
        )
        if not html:
            return url
        return format_html(
            f'<a href="{url}" id="{field}-object-url">{object_id}</a>'
        )
    except NoReverseMatch:
        return object_id
    except AttributeError:
        return '-'
```

:eyes:  Notice the html argument which will allow you to get the URL without the HTML so you don't need to split in `related_object`.

You can then call this method in all the other methods which produce an HTML link. 

Get used to avoid repetition (DRY: Do not repeat yourself, it's a concept also expressed in the book I should have suggested to you).

Reply:
Thanks for suggestion. I see how much it improves quality of code. I try to follow DRY and other principles in future. Should I apply similar logic in tests as well?

Reply:
Regarding tests, it can be also useful but it's less black and white and more a matter of judgement. In general, if I have the same test logic repeated in more than 2 tests, I may try to generalize it in a test method and reuse it to test more cases faster. I just have done that here: https://github.com/openwisp/openwisp-monitoring/pull/48/files#diff-0d2e8c67f37123458a0ec4cbbcc3b9d6L97-R137